### ISSUE TITLE
Extend Operations Store to allow users to pass in a way to cache results

### ISSUE THREAD

**Issue by Austio** [rmosolgo/graphql-ruby#2726]
Howdy, we are investigating using the operation_store for persisted queries in graphql-pro.  

Was wondering if there is a way to allow applications the ability to supply an implementation for the method to `get` from a datastore when looking up an operation.

Ideas could be 
 - providing some generic GraphqlOperationStoreInterface class with read/write that the applications can define 
 - the application supplying some proc object to the OperationStore.  when that proc is provided the `get` would yield the lookup as a block along with the arguments for the client name and operation alias.

**Comment by rmosolgo** [comment]
Hi! Thanks for asking. 

Let me see if I understand: out of the box, the operation store uses ActiveRecord to read and write to database tables. But, you'd like to customize that persistence strategy, is that right? 

Specifically, you mention customizing the _read_ operation. What about writes, would that matter in your situation? 

If I've understood right, I think it could work, although it would need a couple other operations to power the dashboard views, too.

**Comment by Austio** [comment]
Yes but with less scope, to fulfill what we are needing this would be less about customizing the full persistence strategy and more providing a way to supply a caching layer on top of that persistence.  Customizing the full persistence would be fantastic but maybe overkill for what we/community would need.  Open to what other people think.

**Comment by rmosolgo** [comment]
I just released graphql-pro 1.13.0, which accepts an arbitrary `backend_class:`. It requires a _lot_ of methods, so you wouldn't want to implement it from scratch, but you could extend the built-in `ActiveRecordBackend` and override `def get_operation(client_name, operation_alias)` to customize the lookup. 

But honestly, the easiest approach probably doesn't involve the _backend_, you could just customize `OperationStore` to hit a cache _before_ calling to the backend. What I mean is: 

```ruby 
class CachedOperationStore < GraphQL::Pro::OperationStore
  # This method is called at runtime to fetch a record,
  # Override it to add caching, make sure it still returns an OperationRecord.
  def get(client_name:, operation_alias:)
    data = MyCache.fetch("operation/#{client_name}/#{operation_alias}") do 
      operation_record = super 
      if operation_record 
        # convert the Ruby object into something easily cacheable
        {
          "name" => operation_record.name,
          "body" => operation_record.body,
          "digest" => operation_record.digest
        }
      else 
        nil 
      end 
    end 
    if data 
      OperationRecord.new(name: data["name"], digest: data["digest"], body: data["body"], clients_count: nil)
    else 
      nil 
    end 
  end 
end
```

Then: 

```ruby 
class MySchema < GraphQL::Schema 
  # Use a customized subclass of GraphQL::Pro::OperationStore
  use CachedOperationStore 
end 
```

That way, it will use your caching, regardless of the actual backend call. What do you think, want to give it a try?

**Comment by rmosolgo** [comment]
Please let me know if you have any trouble with it!
